### PR TITLE
Locked SOV Contract

### PR DESCRIPTION
**Pull request by powerhousefrank**
LockedSOV Contract General Explanation: https://docs.google.com/document/d/1nUzZSjmliifOPsY5SmSTUf_hCA1l_TIsoY9rPy8y7-Y/

A more detailed update on the explanation will be done soon.

### CODE REVIEW

File: contracts/locked/LockedSOV.sol [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+	 * @param _vestingRegistry The Vesting Registry Address.
+	 * @param _cliff The time period after which the tokens begin to unlock.
+	 * @param _duration The time period after all tokens will have been unlocked.
+	 * @dev IMPORTANT 1: You have to change Vesting Registry if you want to change Duration and/or Cliff.
+	 * IMPORTANT 2: `_cliff` and `_duration` is multiplied by 4 weeks in this function.
+	 */
+	function changeRegistryCliffAndDuration(
+		address _vestingRegistry,
+		uint256 _cliff,
+		uint256 _duration
```
Comment (ororopickpocket) [inline contracts/locked/LockedSOV.sol:174]: same as above

File: contracts/locked/LockedSOV.sol [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	}
+
+	/* Getter or Read Functions */
+
+	/**
+	 * @notice This returns four weeks duration in unix timestamp.
+	 * @return _duration This returns 2419200.
+	 * @dev This is calculated based on 4 * 7 days * 24 hours * 60 minutes * 60 seconds.
+	 */
+	function fourWeeks() public pure returns (uint256 _duration) {
```
Comment (ororopickpocket) [inline contracts/locked/LockedSOV.sol:350]: why not just using a constant?

File: contracts/locked/LockedSOV.sol [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+		_vestingAddress = _createVesting(msg.sender);
+
+		emit VestingCreated(msg.sender, _vestingAddress);
+	}
+
+	/**
+	 * @notice Stakes tokens for a user who already have a vesting created.
+	 * @dev The user should already have a vesting created, else this function will throw error.
+	 */
+	function stakeTokens() external {
```
Comment (ororopickpocket) [inline contracts/locked/LockedSOV.sol:258]: I would find it nice if there would be a function which 
1. withdraws the unlocked tokens and
2. stakes the locked tokens
at once

**Review by ororopickpocket — COMMENTED** [review]
Nice! Would approve it, but waiting for the tests before doing so ;)

File: contracts/locked/LockedSOV.sol [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+		uint256 _duration,
+		address[] memory _admins
+	) public {
+		require(_SOV != address(0), "Invalid SOV Address.");
+		require(_vestingRegistry != address(0), "Vesting registry address is invalid.");
+
+		SOV = IERC20(_SOV);
+		vestingRegistry = VestingRegistry(_vestingRegistry);
+		cliff = _cliff * fourWeeks();
+		duration = _duration * fourWeeks();
```
Comment (eMarchenko) [inline contracts/locked/LockedSOV.sol:129]: `cliff = _cliff * 4 weeks;`

File: contracts/locked/LockedSOV.sol [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+	/**
+	 * @notice Adds SOV to the user balance (Locked and Unlocked Balance based on `_basisPoint`).
+	 * @param _userAddress The user whose locked balance has to be updated with `_sovAmount`.
+	 * @param _sovAmount The amount of SOV to be added to the locked and/or unlocked balance.
+	 * @param _basisPoint The % (in Basis Point)which determines how much will be unlocked immediately.
+	 */
+	function deposit(
+		address _userAddress,
+		uint256 _sovAmount,
+		uint256 _basisPoint
```
Comment (eMarchenko) [inline contracts/locked/LockedSOV.sol:197]: any user can call `LockedSOV.deposit(addr, 1, <inf>)` to immediately unlock all `addr`'s tokens, effectively disabling `lockedBalances` logic
PS: `<inf> == LockedSOV.getLockedBalance(addr) + 1`

**Review by eMarchenko — CHANGES_REQUESTED** [review]
`deposit` function is vulnerable

File: contracts/locked/LockedSOV.sol [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+	 * @param _vestingRegistry The Vesting Registry Address.
+	 * @param _cliff The time period after which the tokens begin to unlock.
+	 * @param _duration The time period after all tokens will have been unlocked.
+	 * @param _admins The list of Admins to be added.
+	 */
+	constructor(
+		address _SOV,
+		address _vestingRegistry,
+		uint256 _cliff,
+		uint256 _duration,
```
[the remark this answers is not in the document]
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol:120]: Done in b3cc958

File: contracts/locked/LockedSOV.sol (hunk 1 above)
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol:175]: Solved.

File: contracts/locked/LockedSOV.sol (hunk 2 above)
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol]: Changed in 540d9bd

File: contracts/locked/LockedSOV.sol (hunk 4 above)
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol]: Changed in 540d9bd

File: contracts/locked/LockedSOV.sol (hunk 3 above)
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol]: Added in a57eae8

File: contracts/locked/LockedSOV.sol (hunk 5 above)
Comment (powerhousefrank) [reply contracts/locked/LockedSOV.sol:199]: Nice catch. Added in e38c7bb

**Review by eMarchenko — APPROVED** [review]

**Review by eMarchenko — APPROVED** [review]

**Review by ororopickpocket — APPROVED** [review]

**Pull request merged**